File: backend/app/models/document.py
from datetime import datetime
from enum import Enum
import uuid

from sqlalchemy import DateTime, Enum as SQLEnum, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class DocumentStatus(str, Enum):
    UPLOADING = "UPLOADING"
    UPLOADED = "UPLOADED"

    PARSING = "PARSING"
    CHUNKING = "CHUNKING"
    EMBEDDING = "EMBEDDING"
    GRAPH_BUILDING = "GRAPH_BUILDING"

    READY = "READY"
    FAILED = "FAILED"


class Document(Base):
    __tablename__ = "documents"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
    )

    filename: Mapped[str] = mapped_column(
        String,
        nullable=False,
    )

    original_filename: Mapped[str] = mapped_column(
        String,
        nullable=False,
    )

    filepath: Mapped[str] = mapped_column(
        String,
        nullable=False,
    )

    mime_type: Mapped[str] = mapped_column(
        String,
        nullable=False,
    )

    file_size: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    status: Mapped[DocumentStatus] = mapped_column(
        SQLEnum(DocumentStatus),
        default=DocumentStatus.UPLOADING,
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )

    chunks: Mapped[list["DocumentChunk"]] = relationship(
    back_populates="document",
    cascade="all, delete-orphan",
    )

File: frontend/types/document.ts
export type DocumentStatus =
  | "UPLOADING"
  | "UPLOADED"
  | "PARSING"
  | "CHUNKING"
  | "EMBEDDING"
  | "GRAPH_BUILDING"
  | "READY"
  | "FAILED";

export interface Document {
  id: string;

  filename: string;

  original_filename: string;

  mime_type: string;

  file_size: number;

  status: DocumentStatus;

  created_at: string;

  updated_at: string;
}

export interface UploadResponse {
  message: string;

  document: Document;
}

export interface DocumentListItem {
  id: string;

  original_filename: string;

  file_size: number;

  status: DocumentStatus;

  created_at: string;
}